cli/lib/kontena/cli/stacks/upgrade_command.rb

```diff
@@ -19,152 +22,211 @@ class UpgradeCommand < Kontena::Command
+          parent_name: data[:parent_name],
+          name: data[:name]
+        )
+        hint_on_validation_notifications(reader.notifications, reader.loader.source)
+        abort_on_validation_errors(reader.errors, reader.loader.source)
       end
-      merged
+
+      set_env_variables(stack_name, current_grid) # restore envs
+
+      Kontena::Cli::Stacks::ChangeResolver.new(old_data, new_data)
     end
 
-    def execute
-      set_env_variables(stack_name, current_grid)
+    def display_report(changes)
+      will = dry_run? ? "would" : "will"
 
-      local = spinner "Parsing #{pastel.cyan(source)}" do
-        normalize_local_data({'stack' => source, 'depends' => skip_dependencies? ? nil : loader.dependencies}, stack_name)
+      puts "SERVICES:"
+      puts "-" * 40
+
+      unless changes.removed_services.empty?
+        puts pastel.yellow("These services #{will} be removed from master:")
+        changes.removed_services.each { |svc| puts pastel.yellow(" - #{svc}") }
+        puts
       end
 
-      remote = spinner "Reading stack #{pastel.cyan(stack_name)} from master" do
-        normalize_master_data(stack_name, true)
+      unless changes.added_services.empty?
+        puts pastel.green("These new services #{will} be created to master:")
+        changes.added_services.each { |svc| puts pastel.green(" - #{svc}") }
+        puts
       end
 
-      merged = merge_data(local, remote)
+      unless changes.upgraded_services.empty?
+        puts pastel.cyan("These services #{will} be upgraded:")
+        changes.upgraded_services.each do |svc|
+          puts pastel.cyan("- #{svc}")
+        end
+        puts
+      end
 
-      removes = merged.keys.select { |k| merged[k][:local].nil? }
+      puts "STACKS:"
+      puts "-" * 40
 
-      unless removes.empty?
+      unless changes.removed_stacks.empty?
+        puts pastel.red("These stacks #{will} be removed because they are no longer depended on:")
+        changes.removed_stacks.each { |stack| puts pastel.red("- #{stack}") }
         puts
-        puts "Stacks to be removed because they are no longer depended on:"
-        removes.each do |r|
-          puts pastel.yellow("- #{r}")
+      end
+
+      unless changes.replaced_stacks.empty?
+        puts pastel.yellow("These stacks #{will} be replaced by other stacks:")
+        changes.replaced_stacks.each do |installed_name, data|
+          puts "- #{pastel.yellow(installed_name)} from #{pastel.cyan(data[:from])} to #{pastel.cyan(data[:to])}"
         end
         puts
-        unless force?
+      end
+
+      unless changes.added_stacks.empty?
+        puts pastel.red("These new stack dependencies #{will} be installed:")
+        changes.added_stacks.each { |stack| puts pastel.red("- #{stack}") }
+        puts
+      end
+
+      unless changes.upgraded_stacks.empty?
+        puts pastel.cyan("These stacks #{will} be upgraded#{' and deployed' if deploy?}:")
```

Comment:
Should this be skipped if it's a solo-stack without any child stacks dependencies? It's maybe extra noise in that case.

```
$ kontena stack upgrade redis test/redis.yaml 
 [done] Reading stack redis from master     
 [done] Parsing test/redis.yaml     
> Analyzing upgrade ..
> Redis version : 3.0
SERVICES:
----------------------------------------
These services will be upgraded:
- redis/redis

STACKS:
----------------------------------------
These stacks will be upgraded and deployed:
- redis


 [done] Upgrading stack redis     
 [done] Triggering deployment of stack redis     
 [done] Waiting for deployment to start     
 [done] Deploying service redis     
```